### ISSUE TITLE
Add reliability to P2P extension

### ISSUE THREAD

**Issue by arthuro555** [4ian/GDevelop#1969]
The peer to peer extension is built on top of WebRTC which uses UDP in the background. It is therefore not reliable. Some users might want something reliable.
Here are potential solutions I could think of:
1. Let the user chose between a reliable or unreliable connection when establishing a connection and use the PeerJs reliability mechanism (bad UX and control of reliability)
2. Create 2 data connections when connecting, one reliable and one unreliable, and use the reliable one if dataloss = false. (Harder to handle errors and disconnection, not sure if WebRTC allows 2 simultaneous connections, potentially worse performance, divides by 2 the amount of players that can connect)
3. Implement our own reliability mechanism that only gets used when dataloss = false (potentially the slowest)

**Comment by 4ian** [comment]
So I don't think there is really a way to ensure a 100% reliable connection (there must be even theorems proving that a 100% reliable network connection is unfeasible :)). In faced paced action game, there is even a point where *you don't want**to receive an old packet that wast lost. It's too old, you've received new information anyway. 

Instead, I think we should go for a high lever abstraction that has mechanisms to recover from data not received in time.
It of course depends on what kind of app/game you're making:
- for a chat, you want this abstraction to ensure that sending a message is retried as long as you've not received an "acknowledgement" from the other peer(s) that it was received. Probably using a protocol that gives you some guarantee is useful.
- for a fast paced game, you can totally survive the loss of a message indicating the position of a player - it will be overriden anyway by the next message.

This would go in the way of the third solution - of course we can make life easier by using a network protocol that gives strong guarantees about message reception, but nevertheless for a robust and reliable game, we need a higher level reliability mechanism I think.

**Comment by arthuro555** [comment]
> **you don't want** to receive an old packet that wast lost.

Yes what I was thinking of is using reliability only when the no dataloss mode is toggled

**Comment by arthuro555** [comment]
🤔 It seems peerjs is reliable and ordered from what I've seen, unlike I believed. That raises more issues though, and peerjs isn't flexible enough to allow changing that. I begin regretting choosing peerjs, I also only now notice that the maintainers aren't that active (they seem to leave the project abandoned every few months for a few months, don't answer the issues, don't review the PRs...), but now that it is out in GDevelop it's a bit too late. 

Though, as the feature is marked as experimental, maybe it is fine to have it change? I think either [this library](https://github.com/feross/simple-peer) which looks more serious or directly using WebRTC would be good, though the problem is that it doesn't have the nice ID system for connecting, both have two big payloads that need to be exchanged. Maybe we could look into hooking up the peer js server with simple-peer, it shouldn't be too hard as it's really just exchanging via a websockets server a payload.

Anyways, peerjs being seemingly reliable and ordered somehow after all, I'm closing this issue.

**Comment by 4ian** [comment]
>  Though, as the feature is marked as experimental, maybe it is fine to have it change? 

Even for experimental stuff, we'll avoid making breaking changes *but* if you judge there is a way to do something way better using another approach/library/abstraction, then it's always possible to: 

- Solution 1: make a new extension based on this approach/library/abstraction.
  - We then move the old extension from "experimental" to "deprecated" - hiding it as much as possible in the editor, and we recommend people to move to the new extension.
  - If it's rarely used, we could even remove the old extension after 6 months or 1 year. This being said, in GDevelop we fight hard for backward compatibility - so we would avoid such a short timeframe (unless we have proof that almost nobody is really using this extension), as even 1 year is not that much for someone working on a game and because that would undermine trust of users in the stability of GDevelop. An exception is if an extension is particularly bad/create lots of confusion/security issues, in which case we could go faster.
- Solution 2: rework the entire extension internally to use something else than peerjs. This will still break the compatibility a bit because users will have to change how they use it, so it really depends how well the new solution/library maps to the existing one. If it doable with just a few changes for the user, it's acceptable (we can add it the changelog, a note indicating what to do to make it work again). If there is too much, then making a new extension might be a better idea.
